ddl/delete_range.go

```diff
@@ -0,0 +1,331 @@
+	ctxPool      *pools.ResourcePool
+	storeSupport bool
+	emulatorCh   chan struct{}
+	keys         []kv.Key
+}
+
+// newDelRangeManager returns a delRangeManager.
+func newDelRangeManager(d *ddl, ctxPool *pools.ResourcePool, supportDelRange bool) delRangeManager {
+	dr := &delRange{
+		d:            d,
+		ctxPool:      ctxPool,
+		storeSupport: supportDelRange,
+	}
+	if !dr.storeSupport {
+		dr.emulatorCh = make(chan struct{}, delBackLog)
+		dr.keys = make([]kv.Key, 0, delBatchSize)
+	}
+	return dr
+}
+
+// addDelRangeJob implements delRangeManager interface.
+func (dr *delRange) addDelRangeJob(job *model.Job) error {
+	resource, err := dr.ctxPool.Get()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer dr.ctxPool.Put(resource)
+	ctx := resource.(context.Context)
+	ctx.GetSessionVars().SetStatusFlag(mysql.ServerStatusAutocommit, true)
+
+	err = insertBgJobIntoDeleteRangeTable(ctx, job)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !dr.storeSupport {
+		dr.emulatorCh <- struct{}{}
+	}
+	log.Infof("[ddl] add job (%d,%s) into delete-range table", job.ID, job.Type.String())
+	return nil
+}
+
+// start implements delRangeManager interface.
+func (dr *delRange) start() {
+	if !dr.storeSupport {
+		dr.d.wait.Add(1)
+		go dr.startEmulator()
+	}
+}
+
+// clear implements delRangeManager interface.
+func (dr *delRange) clear() {
+	log.Infof("[ddl] closing delRange session pool")
+	dr.ctxPool.Close()
+}
+
+// startEmulator is only used for those storage engines which don't support
+// delete-range. The emulator fetches records from gc_delete_range table and
+// deletes all keys in each DelRangeTask.
+func (dr *delRange) startEmulator() {
+	defer dr.d.wait.Done()
+	log.Infof("[ddl] start delRange emulator")
+	for {
+		select {
+		case <-dr.emulatorCh:
+		case <-dr.d.quitCh:
+			return
+		}
+		dr.doDelRangeWork()
+	}
+}
+
+func (dr *delRange) doDelRangeWork() error {
+	resource, err := dr.ctxPool.Get()
+	if err != nil {
+		log.Errorf("[ddl] delRange emulator get session fail: %s", err)
+		return errors.Trace(err)
+	}
+	defer dr.ctxPool.Put(resource)
+	ctx := resource.(context.Context)
+	ctx.GetSessionVars().SetStatusFlag(mysql.ServerStatusAutocommit, true)
+
+	ranges, err := LoadDeleteRanges(ctx, math.MaxInt64)
+	if err != nil {
+		log.Errorf("[dd] delRange emulator load tasks fail: %s", err)
+		return errors.Trace(err)
+	}
+
+	for _, r := range ranges {
+		if err := dr.doTask(r); err != nil {
+			log.Errorf("[ddl] delRange emulator do task fail: %s", err)
+			return errors.Trace(err)
+		}
+		if err := CompleteDeleteRange(ctx, r); err != nil {
+			log.Errorf("[ddl] delRange emulator complete task fail: %s", err)
+			return errors.Trace(err)
+		}
+		log.Infof("[ddl] delRange emulator complete task: (%d, %d)", r.jobID, r.elementID)
+	}
+	return nil
+}
+
+func (dr *delRange) doTask(r DelRangeTask) error {
+	for {
+		finish := true
+		dr.keys = dr.keys[:0]
+		err := kv.RunInNewTxn(dr.d.store, false, func(txn kv.Transaction) error {
```

Comment:
In fact we won't redo those work. We do a seek every times when we need to delete, so If we delete all keys between [1, 100], next time we seek to 1, we will get 101.

I prefer use only ONE `delete from table` rather than several `update table set start key = ...` because that's more simple, and if there are multi tidb do delete-range concurrently, it will be more safe.